src/main/webapp/app/components/data-table/data-table.component.ts

```diff
@@ -0,0 +1,295 @@
+import { Component, Input, OnInit, OnChanges, SimpleChanges, ContentChild, TemplateRef, ViewEncapsulation } from '@angular/core';
+import { debounceTime, distinctUntilChanged, map, tap } from 'rxjs/operators';
+import { Observable } from 'rxjs';
+import { ColumnMode, SortType } from '@swimlane/ngx-datatable';
+import { SortByPipe } from 'app/components/pipes';
+import { compose, filter } from 'lodash/fp';
+import { get } from 'lodash';
+import { BaseEntity } from 'app/shared';
+
+enum SortOrder {
+    ASC = 'asc',
+    DESC = 'desc',
+}
+
+enum SortIcon {
+    NONE = 'sort',
+    ASC = 'sort-up',
+    DESC = 'sort-down',
+}
+
+const SortOrderIcon = {
+    [SortOrder.ASC]: SortIcon.ASC,
+    [SortOrder.DESC]: SortIcon.DESC,
+};
+
+type SortProp = {
+    field: string;
+    order: SortOrder;
+};
+
+const entityToString = (entity: BaseEntity) => entity.id.toString();
+
+@Component({
+    selector: 'jhi-data-table',
+    templateUrl: './data-table.component.html',
+    styleUrls: ['data-table.component.scss'],
+    encapsulation: ViewEncapsulation.None,
+})
+export class DataTableComponent implements OnInit, OnChanges {
+    @ContentChild(TemplateRef, { read: TemplateRef, static: false })
+    templateRef: TemplateRef<any>;
+
+    @Input() isLoading = false;
+    @Input() entityType = 'entity';
+    @Input() resultName = 'result';
+    @Input() allEntities: BaseEntity[] = [];
+    @Input() entitiesPerPageTranslation: string;
+    @Input() searchPlaceholderTranslation: string;
+    @Input() searchFields: string[] = [];
+    @Input() searchTextFromEntity: (entity: BaseEntity) => string = entityToString;
+    @Input() searchResultFormatter: (entity: BaseEntity) => string = entityToString;
+    @Input() customFilterKey: any = {};
+    @Input() customFilter: (entity: BaseEntity) => boolean = () => true;
+
+    readonly PAGING_VALUES = [10, 20, 50, 100, 200, 500, 1000, 2000];
+    DEFAULT_PAGING_VALUE = 50;
+
+    entities: BaseEntity[];
+    entityCriteria: {
+        textSearch: string[];
+        sortProp: SortProp;
+    };
+
+    entitiesPerPage: number;
+    isRendering: boolean;
+
+    constructor(private sortByPipe: SortByPipe) {
+        this.entities = [];
```

Comment:
I moved the bindings into `ngOnInit`. It turned out that `entityCriteria` needs to already be defined in the constructor (or at least before `ngOnInit`) since `ngOnChanges` needs it and is called before `ngOnInit`.